### PR TITLE
Update jupyterlab-ihaskell for 4.x

### PR DESCRIPTION
**Pull request by vaibhavsagar**

### CODE REVIEW

File: jupyterlab-ihaskell/src/index.ts [hunk 1]
```diff
@@ -1,22 +1,29 @@
...
+import { StreamLanguage, LanguageSupport } from '@codemirror/language';
+import {
+  EditorLanguageRegistry,
+  IEditorLanguageRegistry,
+} from '@jupyterlab/codemirror';
 
-import { ICodeMirror } from '@jupyterlab/codemirror';
-
-import { defineIHaskellMode } from './codemirror-ihaskell';
+export const languagePlugin: JupyterFrontEndPlugin<IEditorLanguageRegistry> = {
```
[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply jupyterlab-ihaskell/src/index.ts]: Updated.

File: jupyterlab-ihaskell/src/index.ts [hunk 2]
```diff
@@ -1,22 +1,29 @@
...
+  id: '@jupyterlab/codemirror-extension:language-ihaskell',
+  description: 'A CodeMirror extension for IHaskell',
+  requires: [IEditorLanguageRegistry],
+  activate: (app: JupyterFrontEnd) => {
```
[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply jupyterlab-ihaskell/src/index.ts]: Updated.

File: jupyterlab-ihaskell/src/index.ts [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
+    const languages = new EditorLanguageRegistry();
```
[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply jupyterlab-ihaskell/src/index.ts]: Updated.

File: jupyterlab-ihaskell/src/index.ts [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
+      name: 'ihaskell',
+      mime: 'text/x-ihaskell',
+      load: async () => {
+        const hs = await Promise.resolve(import('@codemirror/legacy-modes/mode/haskell'));
+        const parser = StreamLanguage.define(hs.haskell);
+        const languageSupport = new LanguageSupport(parser);
+        return languageSupport;
+      }
+    });
+    return languages;
```
[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply jupyterlab-ihaskell/src/index.ts]: Updated.

File: jupyterlab-ihaskell/src/index.ts [hunk 5]
```diff
@@ -1,22 +1,29 @@
...
-  autoStart: true,
-  requires: [ICodeMirror],
-  activate: (app: JupyterFrontEnd, codeMirror: ICodeMirror) =>
-  {
-    defineIHaskellMode(codeMirror).catch(console.warn);
+    languages.addLanguage({
+      name: 'ihaskell',
+      mime: 'text/x-ihaskell',
+      load: async () => {
+        const hs = await Promise.resolve(import('@codemirror/legacy-modes/mode/haskell'));
```
[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply jupyterlab-ihaskell/src/index.ts]: Your code (which I based this off) had `Promise.all` with multiple imports, but I only care about one import. I've moved to importing this at the top of my file but that doesn't seem to be making a difference.

**Comment by vaibhavsagar** [comment]
Thanks for your help! I think I've made all the changes you suggested here but I'm still getting a blank screen and exactly the same error message. Any more ideas?

**Comment by erkin98** [comment]
did you fix it @vaibhavsagar ?

**Comment by vaibhavsagar** [comment]
No, this PR doesn't work yet.

**Comment by erkin98** [comment]
> No, this PR doesn't work yet.

how can i reproduce?

**Comment by vaibhavsagar** [comment]
You can run

```
$ nix build -L .#ihaskell-env-ghc94
```

on this branch to build a JupyterLab environment with IHaskell and `jupyterlab-ihaskell` installed. To make changes to `jupyterlab-ihaskell`, follow the instructions in that directory's `README`.

**Comment by vaibhavsagar** [comment]
I need to reinstate the R modes and the `:!` behaviour, but for now this is good enough and already an improvement over the existing lack of syntax highlighting.

**Pull request merged**